plotly/graph_objs/graph_objs.py

```diff
@@ -79,32 +126,96 @@ def __init__(self, *args):
                       ">>> {name}(dict(), dict())"
                       "".format(name=self.__class__.__name__)
             )
-        self.validate()
-        if self.__class__.__name__ == 'PlotlyList':
-            warnings.warn("\nThe PlotlyList class is a base class of "
-                          "list-like graph_objs.\nIt is not meant to be a "
-                          "user interface.")
 
-    def to_graph_objs(self, caller=True):
-        """Change any nested collections to subclasses of PlotlyDict/List.
+        super(PlotlyList, self).__init__()
+
+        for index, value in enumerate(list(*args)):
+            try:
+                value = self.value_to_graph_object(index, value, _raise=_raise)
+            except exceptions.PlotlyGraphObjectError as err:
+                err.prepare()
+                raise
+
+            if isinstance(value, PlotlyBase):
+                self.append(value)
+
+    def __setitem__(self, index, value, _raise=True):
+        """Override to enforce validation."""
+        if not isinstance(index, int):
+            if _raise:
+                index_type = type(index)
+                raise TypeError('Index must be int, not {}'.format(index_type))
+            return
+
+        if index >= len(self):
+            raise IndexError(index)
+
+        value = self.value_to_graph_object(index, value, _raise=_raise)
+        if isinstance(value, (PlotlyDict, PlotlyList)):
+            value.__dict__['_parent'] = self
+            super(PlotlyList, self).__setitem__(index, value)
+
+    def __setattr__(self, key, value):
+        raise exceptions.PlotlyError('Setting attributes on a PlotlyList is '
+                                     'not allowed')
+
+    def __copy__(self):
+
+        # TODO: https://github.com/plotly/python-api/issues/291
+        return GraphObjectFactory.create(self._name, *self)
+
+    def __deepcopy__(self, memodict={}):
+
+        # TODO: https://github.com/plotly/python-api/issues/291
+        return self.__copy__()
 
-        Procedure:
-            1. Attempt to convert all entries to a subclass of PlotlyDict.
-            2. Call `to_graph_objects` on each of these entries.
+    def append(self, value):
+        """Override to enforce validation."""
+        index = len(self)  # used for error messages
+        value = self.value_to_graph_object(index, value)
+        value.__dict__['_parent'] = self
+        super(PlotlyList, self).append(value)
 
+    def extend(self, iterable):
+        """Override to enforce validation."""
+        for value in iterable:
+            index = len(self)
+            value = self.value_to_graph_object(index, value)
+            super(PlotlyList, self).append(value)
+
+    def insert(self, index, value):
+        """Override to enforce validation."""
+        value = self.value_to_graph_object(index, value)
+        value.__dict__['_parent'] = self
+        super(PlotlyList, self).insert(index, value)
```

Comment:
The alternative way to do this would be to make this a `MutableSequence` and only write out `__getitem__`, `__setitem__`, `__delitem__`, `__len__`, and `insert`: https://docs.python.org/2/library/collections.html#collections-abstract-base-classes
 and get `append`, `reverse`, `extend`, `pop`, `remove`, and `__iadd__` for free.